### PR TITLE
Bus stop placement

### PR DESCRIPTION
**Pull request by jenrik**
Buses are placed using the poisson-disc algorithm. It's currently extremely slow, but a better algorithm for picking points that are both on a edge and with in the poisson-disc would fix that.

### CODE REVIEW

File: randomActivityGen.py
Comment (NicEastvillage) [inline randomActivityGen.py]: Good idea to use existing bus stops as seeds. However, `bus_stations.findall("BusStation")` returns a list of xml elements and not points.

File: utility.py
Comment (NicEastvillage) [inline utility.py]: This seems close to what the `map` function does. Doe `map` not support generators?

File: randomActivityGen.py
Comment (NicEastvillage) [inline randomActivityGen.py]: SUMO's units are meters, so this default value means there will be bus stops every 10-20 meters 🤔 
I think something like ~300 is more fitting

**Review by NicEastvillage — CHANGES_REQUESTED** [review]
It's hard to tell how well it works without any visuals, but I assume there has been no major changes to the algorithms results. But I guess, that's one of the next thing on the todo list.

I do have some concerns about a certain default value and some xml, however.

File: utility.py
Comment (jenrik) [reply utility.py]: Fixed in 2f5cbb2. I wasn't aware `map` existed, and searching the internet didn't yield anything for me

File: randomActivityGen.py
Comment (jenrik) [reply randomActivityGen.py]: Changed in 5d6b58d, 10 was just a temporary value I forgot to change

File: bus.py
Comment (falkecarlsen) [inline bus.py:40]: Assertions doesn't mesh with the logging, I found out but maybe we don't care about that.

**Review by falkecarlsen — COMMENTED** [review]
Just one comment. I'm not strong enough in Python to entirely understand how it works nor debug it sufficiently but it's pretty to read and seems well documented :+1:

File: randomActivityGen.py
Comment (jenrik) [reply randomActivityGen.py]: fixed in 16f8af1

File: bus.py
Comment (jenrik) [reply bus.py:40]: An assertion matches your description of critical, but it also halt's execution. If I an assertion is violated then code is defective, there for are there no reason to continue running.

**Comment by jenrik** [comment]
When the default distance where changed from 10 to 500 the run time on the example network changed from more than a hour to a couple of seconds. Maybe performance isn't all that bad, but I would like to do a test on a much larger network before deciding to go forth with optimizing

**Comment by NicEastvillage** [comment]
> from _more than a hour_ to a couple of seconds

Wtf?! I mean, it makes sense. But we definitely can't ship it if it takes an hour to run. The script is meant for fast (but semi-realistic) iteration xD Good thing we got that fixed

File: randomActivityGen.py
Comment (NicEastvillage) [inline randomActivityGen.py]: Use `net.getEdge(edge_id)` instead of iterating through them

File: bus.py
Comment (NicEastvillage) [inline bus.py]: Never use mutable defaults arguments. See https://docs.python-guide.org/writing/gotchas/

File: bus.py
Comment (NicEastvillage) [inline bus.py]: I am pretty sure something is wrong here `len(all_points) <= 0` is only true if `len(all_points) == 0`, but you are doing `all_points.add(road) # Seed point` in the body, so swapping the `<=` to `>=` would result in an infinite loop.

File: bus.py
Comment (NicEastvillage) [inline bus.py:61]: I would personally invert the logic of this or rename it. Currently it sounds like it "checks if point is okay" instead of "checks if point is NOT okay".

**Review by NicEastvillage — CHANGES_REQUESTED** [review]
I found a few things. See below.

Also, please replace your print statements with loggings now that we have it setup.

File: bus.py
Comment (jenrik) [reply bus.py:61]: I disagree, previously I had it reversed but that became to confusing

Comment (NicEastvillage) [reply bus.py:61]: You sure? You are inverting everywhere already.
Example L80-86 in bus.py:
```python
# ...
try:
    # Search for a point, or raise StopIteration is none can be found
    point = next(filter(
            lambda p: not check_dist(p, iter(all_points)),
            gen))

    assert not check_dist(point, all_points)
# ...
```

File: bus.py
Comment (jenrik) [reply bus.py]: Clarified the intend of the code in 00587e0. I use `len(x) <= y` to check for emptiness and cover the rare case that a length function produces a negative value. I tend to code defensively like that.

File: bus.py
Comment (jenrik) [reply bus.py]: fixed in 77cbece

File: bus.py
Comment (NicEastvillage) [reply bus.py]: I see. I get it now. Though typically lengths are not negative 🤔

File: randomActivityGen.py
Comment (jenrik) [reply randomActivityGen.py]: fixed in 434f43f

File: bus.py
Comment (jenrik) [reply bus.py]: I agree length are **typically** not negative, but this covers faulty code and unknown sentinel values

File: bus.py
Comment (jenrik) [reply bus.py:61]: I wrote it the way I found most logical, and easy to read. Is there any problem with function of the code, otherwise we are just pointlessly arguing over wording.

**Review by NicEastvillage — APPROVED** [review]

**Review by MathiasMehl — APPROVED** [review]

**Comment by MathiasMehl** [comment]
Approved, but would be nice to also render the busstops, should be relatively easy. Can be done in another PR afterwards

**Pull request merged**
